pkg/operator/ceph/cluster/mgr/mgr.go

```diff
@@ -44,38 +45,48 @@ const (
 
 // Cluster represents the Rook and environment configuration settings needed to set up Ceph mgrs.
 type Cluster struct {
-	Namespace   string
-	Replicas    int
-	placement   rookalpha.Placement
-	context     *clusterd.Context
-	dataDir     string
-	HostNetwork bool
-	resources   v1.ResourceRequirements
-	ownerRef    metav1.OwnerReference
-	dashboard   cephv1.DashboardSpec
-	cephVersion cephv1.CephVersionSpec
-	rookVersion string
-	exitCode    func(err error) (int, bool)
-	clusterInfo *cephconfig.ClusterInfo
+	cephIdentity *identity.Identity
```

Comment:
The answer is my sort of long-term thinking ... 

I'd like to remove the `ClusterInfo` struct at a later date; I don't particularly like how it is tied to more than one thing at once (mons and config), and I wanted a singularly focused, small thing here. It has mon info in it that is necessary currently, but I hope to make it unnecessary later. 

I kind of wanted a clean slate on a struct for keeping the cluster identity, especially since an important part of what I wanted to achieve was to add some explicitness around the fact that the information starts its life as blank/unknown and is later explicitly expected to exist after the mons have been processed.

Ultimately, the `ClusterInfo` struct could be used, and I think there are good reasons to use it (not reinventing the wheel, etc.) if others have a preference to do so.

Reply:
While it is true that the fsid is not known when the structs are initialized, the very first thing that happens in the mon creation is that the fsid, admin secret, and mon secret are all initialized together into the ClusterInfo struct. After they are created/loaded, they will never change. This is why in the operator code we can require that they are always set. I wouldn't expect that we need to have a lock on them either since they will never be changed after being initially set. Reading from multiple goroutines will be fine as long as they never change.

I could see why you would want to separate the other fields of the ClusterInfo such as the mon endpoints and version name since those could change over time. I'm not seeing why we would want to separate the FSID from the mon and admin secret since they are created at the same time and are all immutable for the cluster. Either we should break them out together or leave them in the ClusterInfo. If you need the fsid elsewhere before a full refactoring of the ClusterInfo, I'd suggest just passing the fsid as a string where needed.

Reply:
I think everything except the FSID *can* change, though. Unless I'm mistaken, the admin key and mon key(s) could be revoked and replaced with a different key. Otherwise, a new Ceph cluster would have to be created if either of those keys were compromised. 

I do see the point that without those keys, Rook can't begin to manage the Ceph cluster, and in that sense, I see that those keys are critical for setting up a new cluster or connecting to an existing one and -- immutable or not -- part of the identity as far as Rook is concerned.

I think I will refactor to use `ClusterInfo`, but I still want to have some high-level indication that it starts its life empty (some sort of `BlankInfo` method) and that it is required (`IsEstablished`) to be non-empty after the mons have been configured so that devs inspecting the code don't have to inspect the entirety of the mon code to understand where its life begins at a high level.

Reply:
Certainly, if it helps the clarity of the code, sounds good to keep the `IsEstablished()` or similar check so it's clear that it's initialized. What about calling it `IsInitialized()`? IsEstablished gives me the impression that it's related to a network connection, but maybe that's just me. 

Since ClusterInfo is passed around most places with a pointer, I'd suggest that `nil` be used to indicate it's blank, but we can see how your changes work out with the `BlankInfo()` if you'd prefer.

While the admin and mon keys could change, Rook doesn't expect it today, so it would be good to track what should be done around this.

Reply:
I ended up starting the `ClusterInfo` as `nil`, which I liked, and then I return the `ClusterInfo` struct from the mons. The `nil` info in the `cluster` is then set with this return value; I like the extra transparency of this. And I did like `IsInitialized` better for verification.